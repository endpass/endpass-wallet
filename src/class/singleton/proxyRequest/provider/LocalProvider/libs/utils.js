import regPath from 'path-to-regexp';

const utils = {
  getRouteKey(routesRegexp, routePath) {
    return Object.keys(routesRegexp).find(key => {
      const re = routesRegexp[key];
      return re.test(routePath);
    });
  },
  getPath(serverUrl, url) {
    const levels = url.split(serverUrl);
    const keysPath = levels.pop() || '';
    const ret = keysPath[0] === '/' ? keysPath.substring(1) : keysPath;

    return ret;
  },
  getPathArgs(routesRegexp, key, routePath) {
    const keys = regPath.parse(key);
    const re = routesRegexp[key];
    re.lastIndex = 0; // drop exec position

    const values = re.exec(routePath);
    return keys.reduce((map, item, index) => {
      if (item && item.name) {
        map[item.name] = values[index];
      }
      return map;
    }, {});
  },
};

export default utils;
