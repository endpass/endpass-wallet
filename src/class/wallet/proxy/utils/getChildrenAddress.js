export default function(hdWallet, offset, limit) {
  return [...Array(limit)].map((_, i) =>
    hdWallet
      .deriveChild(offset + i)
      .getWallet()
      .getChecksumAddressString(),
  );
}
