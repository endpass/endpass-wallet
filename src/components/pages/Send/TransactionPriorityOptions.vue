<template>
  <div class="priority-options field is-horizontal">
    <div class="field-label">
      <label class="label">Priority</label>
    </div>
    <div class="field-body">
      <v-radio
        v-if="suggestedGasPrices"
        id="priority"
        :options="suggestedGasPrices"
        :value="value"
        name="priority"
        @input="emitPriorityInput"
      />
      <v-spinner v-else-if="isLoading" />
      <p
        v-else
        class="help is-danger"
      >
        Unable to load suggested gas price, please set gas price
        manually.
      </p>
    </div>
  </div>
</template>

<script>
import VRadio from '@/components/ui/form/VRadio';

export default {
  name: 'TransactionPriorityOptions',

  props: {
    isLoading: {
      type: Boolean,
      default: false,
    },

    prices: {
      type: Object,
      default: () => ({}),
    },

    value: {
      type: String,
      required: true,
    },
  },

  computed: {
    suggestedGasPrices() {
      const { prices } = this;

      if (prices) {
        return [
          {
            val: prices.low.toString(),
            key: 'Low',
            help: `${prices.low} Gwei`,
          },
          {
            val: prices.medium.toString(),
            key: 'Medium',
            help: `${prices.medium} Gwei`,
          },
          {
            val: prices.high.toString(),
            key: 'High',
            help: `${prices.high} Gwei`,
          },
        ];
      }

      return null;
    },
  },

  methods: {
    emitPriorityInput(value) {
      this.$emit('input', value);
    },
  },

  components: {
    VRadio,
  },
};
</script>
