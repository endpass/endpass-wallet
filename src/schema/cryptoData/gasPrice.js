export const cryptoDataGasPrice = {
  type: 'object',
  additionalProperties: false,
  required: ['low', 'medium', 'high'],
  properties: {
    low: {
      type: 'number',
    },
    medium: {
      type: 'number',
    },
    high: {
      type: 'number',
    },
  },
};
