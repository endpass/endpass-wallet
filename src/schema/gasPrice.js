import { ajv } from '@/class/singleton';

export const gasPriceSchema = {
  properties: {
    low: {
      type: 'number',
      default: 0,
    },
    medium: {
      type: 'number',
      default: 0,
    },
    high: {
      type: 'number',
      default: 0,
    },
  },
};

export default {
  validateGasPrice: ajv.compile(gasPriceSchema),
};
