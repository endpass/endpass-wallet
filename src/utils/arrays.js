export const uniq = arr =>
  arr.reduce((acc, item) => (acc.includes(item) ? acc : acc.concat(item)), []);

export const mapArrayByProp = (arr, prop) =>
  arr.reduce((acc, item) => {
    const target = item[prop];

    if (target) {
      return Object.assign(acc, {
        [target]: item,
      });
    }

    return acc;
  }, {});

export default {
  uniq,
  mapArrayByProp,
};
