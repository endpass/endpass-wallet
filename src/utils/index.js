import store from '../store';

export { default as keystore } from './keystore';
export { default as proxyRequest } from './proxyRequest';

export const getInitializedValueFromStore = (module, field) =>
  new Promise(resolve => {
    const value = module[field];
    if (value === null) {
      const unwatch = store.watch(
        () => module[field],
        newValue => {
          unwatch();
          resolve(newValue);
        },
      );
    } else {
      resolve(value);
    }
  });
