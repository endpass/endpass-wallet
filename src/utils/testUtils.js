import { kebabCase } from 'lodash';
import { getFrom } from '@/utils/objects';

function convertDirectivesToAttrs(directives) {
  const attrs = {};

  if (directives) {
    directives.forEach(({ rawName, expression }) => {
      attrs[rawName] = expression;
    });
  }

  return attrs;
}

function convertListenersToAttrs(listeners) {
  const attrs = {};

  if (listeners) {
    Object.keys(listeners).forEach(key => {
      const handlers = [].concat(listeners[key]);
      const handler = handlers.find(h => {
        const name = getFrom(h, 'fns.name', 'name') || '';

        return name.includes('bound ');
      });

      if (handler) {
        attrs[`v-on:${key}`] = (handler.fns || handler).name;
      }
    });
  }

  return attrs;
}

function convertModelToAttrs(model) {
  return model ? { 'v-model': model.expression } : {};
}

function convertSlotToAttrs(slot) {
  return slot ? { slot } : {};
}

function generateElement(createElement, vnode) {
  return vnode.componentOptions
    ? generateElementFromComponent(createElement, vnode)
    : generateElementFromHTML(createElement, vnode);
}

function getChildrenElements(createElement, children) {
  return children
    ? children.map(child => generateElement(createElement, child))
    : undefined;
}

function generateElementFromComponent(
  createElement,
  { data, componentOptions },
) {
  const { attrs, directives, model } = data;
  const { tag, listeners, children } = componentOptions;
  const componentData = {
    ...data,
    attrs: {
      ...attrs,
      ...convertDirectivesToAttrs(directives),
      ...convertListenersToAttrs(listeners),
      ...convertModelToAttrs(model),
      class: data.staticClass,
    },
    hook: undefined,
    staticClass: undefined,
  };

  if (listeners) {
    componentData.on = Object.assign({}, listeners);
  }

  return createElement(
    tag,
    componentData,
    getChildrenElements(createElement, children),
  );
}

function generateElementFromHTML(createElement, vnode) {
  if (!vnode.data) {
    return vnode;
  }

  const { tag, data, children } = vnode;
  const { directives, on, model, attrs = {}, slot } = data;

  return createElement(
    tag,
    {
      ...data,
      attrs: {
        ...attrs,
        ...convertDirectivesToAttrs(directives),
        ...convertListenersToAttrs(on),
        ...convertModelToAttrs(model),
        ...convertSlotToAttrs(slot),
      },
    },
    getChildrenElements(createElement, children),
  );
}

export function generateStubs(Component) {
  return Object.values(Component.components).reduce((stubs, stubComponent) => {
    let elementName;

    if (stubComponent.name) {
      elementName = kebabCase(stubComponent.name);
    }

    return Object.assign(stubs, {
      [elementName]: {
        render(createElement) {
          return generateElementFromComponent(createElement, this.$vnode);
        },
      },
    });
  }, {});
}

export default {
  generateStubs,
};
