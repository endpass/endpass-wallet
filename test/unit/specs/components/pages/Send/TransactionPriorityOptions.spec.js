import VeeValidate from 'vee-validate';
import { shallow, createLocalVue } from '@vue/test-utils';
import { generateStubs } from '@/utils/testUtils';
import TransactionPriorityOptions from '@/components/pages/Send/TransactionPriorityOptions.vue';
import { gasPrice } from 'fixtures/gasPrice';

const localVue = createLocalVue();

localVue.use(VeeValidate);

describe('Send â€“ TransactionPriorityOptions', () => {
  const mountProps = {
    prices: gasPrice,
    value: gasPrice.medium.toString(),
  };
  let wrapper;

  beforeEach(() => {
    wrapper = shallow(TransactionPriorityOptions, {
      provide: () => ({
        $validator: new VeeValidate.Validator(),
      }),
      stubs: generateStubs(TransactionPriorityOptions),
      propsData: mountProps,
      localVue,
    });
  });

  describe('render', () => {
    it('should render correctly prices', () => {
      expect(wrapper.html()).toMatchSnapshot();
    });

    it('should not render options if prices are empty', () => {
      wrapper = shallow(TransactionPriorityOptions, {
        provide: () => ({
          $validator: new VeeValidate.Validator(),
        }),
        stubs: generateStubs(TransactionPriorityOptions),
        propsData: {
          prices: null,
          value: '0',
        },
        localVue,
      });

      expect(wrapper.html()).toMatchSnapshot();
    });
  });
});
