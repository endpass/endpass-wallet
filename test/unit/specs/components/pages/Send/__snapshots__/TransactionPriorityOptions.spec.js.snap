// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Send – TransactionPriorityOptions render should not render options if prices are empty 1`] = `
<div class="priority-options field is-horizontal">
    <div class="field-label">
        <label class="label">Priority</label>
    </div>
    <div class="field-body">
        <p class="help is-danger">
            Unable to load suggested gas price, please set gas price manually.
        </p>
    </div>
</div>
`;

exports[`Send – TransactionPriorityOptions render should render correctly prices 1`] = `
<div class="priority-options field is-horizontal">
    <div class="field-label">
        <label class="label">Priority</label>
    </div>
    <div class="field-body">
        <v-radio id="priority" options="[object Object],[object Object],[object Object]" value="3" name="priority" v-on:input="bound emitPriorityInput"></v-radio>
    </div>
</div>
`;
